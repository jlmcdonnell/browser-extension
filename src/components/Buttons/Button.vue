<script lang="ts" setup>
import { computed, toRefs } from 'vue';

type Props = {
  color?: string;
  textColor?: string;
  href?: string;
  size?: 'small' | 'medium' | 'large';
};

const props = defineProps<Props>();
const type = computed(() => {
  if (props.href) {
    return 'a';
  }
  return 'button';
});

const { color, textColor } = toRefs(props);
const classes = computed(() => {
  let colors = '';
  let sizeClass = '';

  if (color?.value) {
    colors = `bg-${color.value} text-${textColor?.value || 'white'}`;
  }

  if (props.size) {
    sizeClass = `btn-${props.size}`;
  }

  return `${colors} ${sizeClass}`;
});
</script>

<template>
  <component :is="type" :href="href" class="btn" :class="classes">
    <slot></slot>
  </component>
</template>

<style scoped>
.btn {
  --green: #44ad4de6;
  --green-hover: #44ad4dff;
  --red: #e34039e6;
  --red-hover: #e34039ff;
  --blue: rgb(41 77 115 / 90%);
  --blue-hover: rgb(41 77 115 / 100%);

  background-color: var(--blue);
  padding: 0.5rem 1rem;
  color: #fff;
  border-radius: 0.25rem;
}

.btn:focus,
.btn:active {
  outline: none;
}

.btn:not(:disabled):focus,
.btn:not(:disabled):hover {
  background-color: var(--blue-hover);
}

.n-button-group .btn:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.n-button-group .btn:not(:last-child) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.n-button-group .btn:not(:first-child, :last-child) {
  border-radius: 0;
}

.bg-success {
  background-color: var(--green);
}

.bg-error {
  background-color: var(--red);
}

.bg-success:not(:disabled):focus,
.bg-success:not(:disabled):hover {
  background-color: var(--green-hover);
}

.bg-error:not(:disabled):focus,
.bg-error:not(:disabled):hover {
  background-color: var(--red-hover);
}

.btn-small {
  padding: 0.25rem 0.5rem;
  font-size: 0.875rem;
}

.btn-medium {
  padding: 0.5rem 1rem;
  font-size: 1rem;
}

.btn-large {
  padding: 0.75rem 1.5rem;
  font-size: 1.125rem;
}
</style>
