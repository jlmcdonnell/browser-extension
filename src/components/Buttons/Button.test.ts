import { mount } from '@vue/test-utils';

import Button from '@/components/Buttons/Button.vue';

describe('Button', () => {
  it('should render an a tag', () => {
    const wrapper = mount(Button, {
      props: { href: 'https://www.mullvad.net' },
      slots: { default: 'Mullvad' },
    });
    const aTag = wrapper.find('a');
    expect(aTag.exists()).toBe(true);
    expect(aTag.attributes('href')).toBe('https://www.mullvad.net');

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render a button', () => {
    const wrapper = mount(Button, {
      slots: { default: 'Knapp' },
    });
    const button = wrapper.find('button');
    expect(button.exists()).toBe(true);

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render a blue button', () => {
    const wrapper = mount(Button, {
      props: { color: 'blue' },
      slots: { default: 'Knapp' },
    });
    const button = wrapper.find('button');
    expect(button.exists()).toBe(true);
    expect(button.classes().includes('bg-blue')).toBe(true);

    expect(wrapper.element).toMatchSnapshot();
  });
});
