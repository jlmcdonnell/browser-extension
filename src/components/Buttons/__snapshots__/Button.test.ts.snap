// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render a blue button 1`] = `
<button
  class="btn bg-blue text-white"
>
  
  Knapp
  
</button>
`;

exports[`Button should render a button 1`] = `
<button
  class="btn"
>
  
  Knapp
  
</button>
`;

exports[`Button should render an a tag 1`] = `
<a
  class="btn"
  href="https://www.mullvad.net"
>
  
  Mullvad
  
</a>
`;
